x-pack/plugins/enterprise_search/server/lib/engines/field_capabilities.ts

```diff
@@ -22,13 +25,134 @@ export const fetchEngineFieldCapabilities = async (
     include_unmapped: true,
     index: getEngineIndexAliasName(name),
   });
+  const fields = parseFieldsCapabilities(fieldCapabilities);
   return {
     created,
     field_capabilities: fieldCapabilities,
+    fields,
     name,
     updated,
   };
 };
 
 // Note: This will likely need to be modified when engines move to es module
 const getEngineIndexAliasName = (engineName: string): string => `search-engine-${engineName}`;
+
+export const parseFieldsCapabilities = (fieldCaps: FieldCapsResponse): SchemaField[] => {
+  const { fields } = fieldCaps;
+  return Object.entries<Record<string, FieldCapsFieldCapability>>(fields)
+    .filter(([fieldName]) => isTopLevelField(fieldName))
+    .map(([fieldName, fieldValue]) => ({
+      ...parseFieldCapability(
+        fieldValue,
+        fieldCaps.indices,
+        getSubFieldCapabilities(fieldName, fields)
+      ),
+      name: fieldName,
+    }));
+};
+
+export const isTopLevelField = (fieldName: string): boolean => {
+  if (fieldName.startsWith('_')) return false;
+  if (fieldName.includes('.')) return false;
+  return true;
+};
+
+export const parseFieldCapability = (
+  value: Record<string, FieldCapsFieldCapability>,
+  indices: string | string[],
+  subFields: SubField[]
+): { fields?: SchemaField[]; indices: SchemaFieldIndex[]; type: string } => {
+  const typeKeys = Object.keys(value);
+  if (typeKeys.length === 0) {
+    return {
+      indices: indicesWithMatchingType('unknown', indices),
+      type: 'unknown',
+    };
+  }
+  let fields: SchemaField[] | undefined;
+  if (subFields.length > 0) {
+    fields = subFields
+      .filter(({ path }) => path.length === 1)
+      .map((subfield) => {
+        return {
+          ...parseFieldCapability(
+            subfield.capability,
+            indices,
+            getSubFields(subfield.path[0], subFields)
+          ),
+          name: subfield.path[0],
+        };
+      });
+  }
+  if (typeKeys.length === 1) {
+    return {
+      fields,
+      indices: indicesWithMatchingType(value[typeKeys[0]].type, indices),
+      type: value[typeKeys[0]].type,
+    };
+  }
+  let type: string;
+  if ('unmapped' in value && typeKeys.length === 2) {
+    type = typeKeys.filter((v) => v !== 'unmapped')[0];
+  } else {
+    type = 'conflict';
+  }
+  const fieldIndices: SchemaFieldIndex[] = [];
```

Comment:
nitpick(non-blocking): for consistency we generally prefer solving this with a `reduce` instead of a for loop (though yes, that's less performant):
```
const fieldIndices = Object.values<FieldCapsFieldCapability>(value).filter(val => !!val.indices)
    .reduce((prev, curr) => [...prev, ...indicesWithMatchingType(curr.type, curr.indices)], []);
```

Reply:
went with a `flatMap`: https://github.com/elastic/kibana/pull/152953/commits/2a3ca4d99cfcdf338a9042ec8f60a890ce58771a